debian-bo-master/bo/source/sound/xmp-0.99c.orig/xmp.1
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH XMP 1 "Version 0.99a" "13/Mar/97" " "
.IX Title "XMP 1"
.UC
.IX Name "xmp - extended module player"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
xmp \- extended module player
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
\fBxmp\fR
[\fB\-8ADhiLlmnorVvx\fR]
[\fB\-d\fR \fIdevice\fR]
[\fB\-M\fR \fIch\fR]
[\fB\-p\fR \fIpan\fR]
[\fB\-S\fR \fIch\fR]
[\fB\-s\fR \fIpos\fR]
\fImodules\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\fBxmp\fR is a module player originally designed to play Fasttracker II
extended modules. The current version recognizes Fasttracker II modules
(XM), 1 to 32 channel Protracker modules (MOD) and variants (including
Mod's Grave WOW modules), Scream Tracker 3 modules (S3M), Scream Tracker 2
modules (STM), Composer 669 and UNIS669 modules (669), Poly Tracker modules
(PTM), Oktalyzer modules (OKT) and Multitracker modules (MTM). Currently
supported output devices are the Gravis Ultrasound (or SoftOSS) and the
Sound Blaster AWE\-32. Software mixer support is NOT implemented in this
version.
.SH "OPTIONS"
.IX Header "OPTIONS"
.Ip "\fB\-8\fR" 8
.IX Item "\fB\-8\fR"
Convert 16-bit samples to 8-bit (to save memory in the sound device).
.Ip "\fB\-A\fR" 8
.IX Item "\fB\-A\fR"
Use Amiga (exponential) period mode. This is the default mode.
.Ip "\fB\-D\fR" 8
.IX Item "\fB\-D\fR"
Dump the pattern data to stdout (for debugging).
.Ip "\fB\-d\fR \fIdevice\fR" 8
.IX Item "\fB\-d\fR \fIdevice\fR"
Uses the specified device for output overriding autodetection. Use this
option if you have more than one output device in your system (e.g. force
softmix in a \s-1GUS\s0). If \fB\-d\fR is not used, xmp will use the first available
device for output. Devices are probed in the order shown in the device
list.
.Ip "\fB\-h\fR" 8
.IX Item "\fB\-h\fR"
Show a short summary of command-line options.
.Ip "\fB\-i\fR" 8
.IX Item "\fB\-i\fR"
Ignore 0xff (end of module marker) when playing S3M files. Not very useful
unless you're playing music from the Unreal demo.
.Ip "\fB\-L\fR" 8
.IX Item "\fB\-L\fR"
Use linear period mode, according to the \s-1XM\s0 specs (period = 7680 \- note<<6
\- finetune>>1).
.Ip "\fB\-l\fR" 8
.IX Item "\fB\-l\fR"
Enable module looping.
.Ip "\fB\-M\fR ch" 8
.IX Item "\fB\-M\fR ch"
Mute the specified channel.
.Ip "\fB\-m\fR" 8
.IX Item "\fB\-m\fR"
Force mono output.
.Ip "\fB\-n\fR" 8
.IX Item "\fB\-n\fR"
Force \s-1NTSC\s0 (60 Hz) timing for \s-1MOD\s0 files. Default is \s-1PAL\s0 (50 Hz).
.Ip "\fB\-p\fR \fInum\fR" 8
.IX Item "\fB\-p\fR \fInum\fR"
Set the percentual panning amplitude. The default value is 75%.
.Ip "\fB\-r\fR" 8
.IX Item "\fB\-r\fR"
Reverse left/right stereo output channels.
.Ip "\fB\-S\fR ch" 8
.IX Item "\fB\-S\fR ch"
Play only the specified channel (solo).
.Ip "\fB\-s\fR \fIpos\fR" 8
.IX Item "\fB\-s\fR \fIpos\fR"
Start playing the module from the position \fIpos\fR.
.Ip "\fB\-V\fR" 8
.IX Item "\fB\-V\fR"
Print version information.
.Ip "\fB\-v\fR" 8
.IX Item "\fB\-v\fR"
Verbose mode (incremental).
.SH "SIGNALS"
.IX Header "SIGNALS"
If \fBxmp\fR receives SIGABRT, it will abort the currently playing module and
start the next module specified in the command line. If SIGUSR1 is received,
\fBxmp\fR will reload the currently playing module.
.SH "CAVEATS"
.IX Header "CAVEATS"
See the file \fBREADME.formats\fR in the distribution package for a complete
list of restrictions for each supported module format.
.SH "BUGS"
.IX Header "BUGS"
.Ip "\(bu" 4
.IX Item "\(bu"
Note slide effects not supported.
.Ip "\(bu" 4
.IX Item "\(bu"
\s-1XM\s0 effects K, L and T not supported.
.Ip "\(bu" 4
.IX Item "\(bu"
\s-1UNIS\s0 669 extended effects not supported.
.Ip "\(bu" 4
.IX Item "\(bu"
\s-1FAR\s0 support is poor.
.PP
See the file \fB\s-1README\s0.bugs\fR in the distribution package for a detailed
list of bugs in the current release. 
.SH "AUTHOR"
.IX Header "AUTHOR"
\fBxmp\fR was written by C Matsuoka <claudio@cpgei.cefetpr.br> and H Carraro Jr
(no current email address). AWE/awedrv support added by Takashi Iwai
<iwai@dragon.mm.t.u-tokyo.ac.jp>.
.SH "DISCLAIMER"
.IX Header "DISCLAIMER"
This is a free product with no warranty of any kind. Use it at your own risk.

.rn }` ''


debian-bo-master/bo/source/sound/xmp-0.99c.orig/xmp.c
/* xmp.c - Fastracker II extended module player
 * Copyright (C) 1996 C.Matsuoka and H.Carraro Jr
 *
 * $Id: xmp.c,v 1.6 1997/03/20 13:19:52 claudio Exp $
 *
 * $Log: xmp.c,v $
 * Revision 1.6  1997/03/20 13:19:52  claudio
 * dump_pattern() fixed to recognize patterns with 256 rows.
 *
 * Revision 1.5  1997/03/18 16:07:45  claudio
 * Added secondary effect to dump_pattern().
 *
 * Revision 1.4  1997/03/13 13:21:26  claudio
 * Lots of changes for xmp 0.99.
 *
 * Revision 1.3  1997/01/19 01:43:09  claudio
 * Added incremental verbosity and ignoreff options.
 *
 * Revision 1.2  1997/01/02 00:19:59  claudio
 * Added options -o and -n.
 *
 * Revision 1.1  1996/12/07 20:40:26  claudio
 * Initial revision
 *
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <time.h>

#include "xmp.h"
#include "xxm.h"
#include "devices.h"

#define TEMP "/tmp/uncompressed.xmp"

extern char *optarg;
extern int optind;
extern char *ver_id;

struct dev_info *d;
struct fmt_info *fmt,*fmt_head;
char module_name[MODULE_NAME_MAXSIZE];
struct options opt;
double replay_rate;
int c4_rate;
int vol_base;
int *vol_xlat;
int p_channels;
struct xmp_channel xc[L_CHANNELS];

struct xxm_header *xxh;
struct xxm_pattern **xxp;		/* Patterns */
struct xxm_track **xxt;			/* Tracks */
struct xxm_channel xxc[32];		/* Channel info */
uint8 xxo[256];				/* Orders */
struct xxm_instrument_header *xxih;	/* Instrument headers */
struct xxm_instrument_map *xxim;	/* Instrument map */
struct xxm_instrument **xxi;		/* Instruments */
struct xxm_sample *xxs;			/* Samples */
uint16 **xxae;				/* Amplitude envelope */
uint16 **xxpe;				/* Pan envelope */
uint16 **xxfe;				/* Pitch envelope */


void dump_pattern(int i)
{
    int r,c,x;
    struct xxm_event *xxe;
    static char *note[12] = {
	"C-","C#","D-","D#","E-","F-",
	"F#","G-","G#","A-","A#","B-"
    };

    x=(uint8)(xxp[i]->rows-1)+1;
    printf("PATTERN %d (%d rows)\n     ",i,x);
    for(c=0;c<xxh->chn;c++)
	printf("[   CH %2d: %03x   ] ",c,xxp[i]->info[c].index);
    printf("\n");
    for(r=0;r<x;r++) {
	printf("[%02x] ",r);
	for(c=0;c<xxh->chn;c++) {
	    xxe=&xxt[xxp[i]->info[c].index]->event[r];
	    if(xxe->not==0x61)
		printf("___ ");
	    else if(xxe->not)
		printf("%s%c ",note[(xxe->not-1)%12],'0'+(xxe->not-1)/12);
	    else printf("... ");
	    if(xxe->ins)
		printf("%02X ",xxe->ins);
	    else printf(".. ");
	    if(xxe->vol)
		printf("%02X ",xxe->vol);
	    else printf(".. ");
	    printf("%c%02X ",xxe->fxt<=9?'0'+xxe->fxt:
		'A'-10+xxe->fxt,xxe->fxp);
	    printf("%c%02X |",xxe->f2t<=9?'0'+xxe->f2t:
		'A'-10+xxe->f2t,xxe->f2p);
	}
	printf("\n");
    }
    printf("\n");
}


void usage()
{
    report(
	"Usage: xmp [-8ADhiLlmnorVv] [-p pan] [-s num] [-d device] [-M ch]\n"
	"	   [-S ch] filename\n"
    );
    report("Available output devices:");
    list_devices(" %s");
    report("\n"
	"    -8      Convert samples to 8 bit\n"
	"    -A      Use Amiga frequency table\n"
#if 0
	"    -B bits Software mixer resolution\n"
#endif
	"    -D      Dump pattern data\n"
#if 0
	"    -F Hz   Software mixer output frequency\n"
#endif
	"    -h      Print this help message\n"
	"    -i      Ignore 0xff patterns in S3M files\n"
	"    -L      Use linear frequency table\n"
	"    -l      Enable module looping\n"
	"    -M ch   Mute channel\n"
	"    -m      Force mono mode\n"
	"    -n      Use NTSC timing in MOD files\n"
	"    -o      Force 3 octave range in MOD files\n"
	"    -p pan  Set panning amplitude to num%%\n"
	"    -r      Reverse left/right channels\n"
	"    -S ch   Set channel to solo mode\n"
	"    -s num  Start from the specified order\n"
	"    -V      Print version information\n"
	"    -v      Verbose mode (incremental)\n"
    );
}


void register_fmt(char *id,int (*loader)())
{
    struct fmt_info *f;

    f=malloc(sizeof(struct fmt_info));
    f->id=malloc(strlen(id)+1);
    strcpy(f->id,id);
    f->loader=loader;
    if(!fmt_head) fmt_head=f;
    else {
	struct fmt_info *i;
	for(i=fmt_head;i->next;i=i->next);
	i->next=f;
    }
    f->next=NULL;
}

void list_fmt(char *s)
{
    struct fmt_info *f;

    for(f=fmt_head;f;f=f->next)
	report(s,f->id);
}

int main(int argc,char **argv)
{
    FILE *f;
    int i,o,mem;
    char module[80],*dev_id=NULL;
    time_t t0,t1;
    int t;

    check_alignment();
#ifdef GUS_DEVICE
    register_device(GUS,"gus","Gravis Ultrasound",32);
#endif
#ifdef AWE_DEVICE
    register_device(AWE,"awe","Sound Blaster AWE-32",32);
#endif
#if 0
    register_device(SOFTMIX,"softmix","Software mixer",32);
#endif

#ifdef LOAD_XM
    register_fmt("Fast Tracker II extended modules (XM)",xm_load);
#endif
#ifdef LOAD_S3M
    register_fmt("Scream Tracker 3 modules (S3M)",s3m_load);
#endif
#ifdef LOAD_STM
    register_fmt("Scream Tracker 2 modules (STM)",stm_load);
#endif
#ifdef LOAD_PTM
    register_fmt("Polytracker modules (PTM)",ptm_load);
#endif
#ifdef LOAD_IT
    register_fmt("Impulse Tracker modules (IT)",it_load);
#endif
#ifdef LOAD_OKT
    register_fmt("Oktalyzer modules (OKT)",okt_load);
#endif
#ifdef LOAD_FAR
    register_fmt("Farandole Composer modules (FAR)",far_load);
#endif
#ifdef LOAD_MTM
    register_fmt("Multitracker modules (MTM)",mtm_load);
#endif
#ifdef LOAD_669
    register_fmt("Composer 669 modules (669)",ssn_load);
#endif
#ifdef LOAD_MOD
    register_fmt("Protracker modules and variants (MOD)",mod_load);
#endif

    /* Defaults */
    opt.reverse=1;
    opt.start=0;
    opt.mix=75;
    i=0;

    while((o=getopt(argc,argv,"8ADd:hiLlM:mnop:rVvS:s:"))!=-1) {
	switch(o) {
	case '8':
	    opt.smp8bit=1;
	    break;
	case 'A':
	    opt.period_mode=XMP_AMIGA_PERIOD_MODE;
	    break;
	case 'D':
	    opt.dump=1;
	    break;
	case 'd':
	    dev_id=malloc(strlen(optarg)); 
	    strcpy(dev_id,optarg);
	    break;
	case 'i':
	    opt.ignoreff=1;
	    break;
	case 'L':
	    opt.period_mode=XMP_LINEAR_PERIOD_MODE;
	    break;
	case 'l':
	    opt.loop=1;
	    break;
	case 'M':
	    if(atoi(optarg)<L_CHANNELS)
		xc[atoi(optarg)].mute=1;
	    break;
	case 'm':
	    opt.reverse=0;
	    break;
	case 'n':
	    opt.ntsc=1;
	    break;
	case 'o':
	    opt.modrange=1;
	    break;
	case 'p':
	    opt.mix=atoi(optarg);
	    if(opt.mix<0) opt.mix=0;
	    if(opt.mix>100) opt.mix=100;
	    break;
	case 'r':
	    opt.reverse=-1;
	    break;
	case 'S':
	    if(atoi(optarg)<L_CHANNELS) {
		for(;i<L_CHANNELS;i++) xc[i].mute=1;
		xc[atoi(optarg)].mute=0;
	    }
	    break;
	case 's':
	    opt.start=atoi(optarg);
	    break;
	case 'V':
	    report("%s\n",ver_id);
	    report("Available output devices:\n");
	    list_devices("\t%s\t\t%s\n");
	    report("Supported module formats:\n");
	    list_fmt("\t%s\n");
	    exit(-1);
	case 'v':
	    opt.verbose++;
	    break;
	case 'x':
	    opt.panel=1;
	    break;
	case 'h':
	    usage();
	default:
	    exit(-1);
	} 
    }
    
    if(!argv[optind]) {
	usage();
	exit(-1);
    }

    d=get_device(dev_id);
    seq_clearmem(d);

    if(opt.verbose>2) {
	report("Output device  : %s \n",d->name);
	if((mem=seq_getfreemem(d))!=-1)
	    report("Available mem  : %d Kb\n",mem/1024);
    }

    xxh=calloc(sizeof(struct xxm_header),1);
    p_channels=d->pch;

    while(optind<argc) {
	strcpy(module,argv[optind]);

	if(module[strlen(module)-1]=='z') {
	    char line[255];
	    sprintf(line,"gzip -dc %s > %s",module,TEMP);
	    if(opt.verbose>2)
		report("Decompression  : %s\n",line);
	    if(system(line)) {
		report("%s: Can't gunzip file.\n",module);
		optind++;
		continue;
	    }
	    strcpy(module,TEMP);
	}    
    
	if((f=fopen(module,"r"))==NULL) {
	    char line[255];
	    sprintf(line,"%s: %s",*argv,module);
	    perror(line);
	    optind++;
	    continue;
	}
    
	seq_clearmem(d);
	replay_rate=PAL_RATE;
	c4_rate=C4_PAL_RATE;
	vol_base=0x40;
	vol_xlat=NULL;

	for(fmt=fmt_head;fmt;fmt=fmt->next)
	    if((i=!fmt->loader(f,d))) break;
	fclose(f);

	unlink(TEMP);

	if(!i) {
	    report("%s: Unrecognized file format\n",module);
	    optind++;
	    continue;
	}

	if(opt.dump) {
	    for(i=0;i<xxh->pat;i++)
		dump_pattern(i);
	    exit(1);
	}

	if((opt.verbose>1)&&((mem=seq_getfreemem(d))!=-1))
	    report("Available mem  : %d Kb\n",mem/1024);

	str_adj(module_name);
	if(!*module_name) strcpy(module_name,"(untitled)");
 
	if(opt.period_mode==XMP_AMIGA_PERIOD_MODE)
	    xxh->flg&=~XXM_FLG_LINEAR;
	if(opt.period_mode==XMP_LINEAR_PERIOD_MODE)
	    xxh->flg|=XXM_FLG_LINEAR;

	if(opt.verbose>1) {
	    if(xxh->flg&XXM_FLG_MODRNG) report("Using Amiga period limits\n");
	    report("Module looping : %s\n",opt.loop?"yes":"no");
	    report("Period mode    : %s\n",
		xxh->flg&XXM_FLG_LINEAR?"linear":"Amiga");
	    report("Restart pos    : %d\n",xxh->rst);
	}
	if(opt.verbose>2) {
	    report("Base volume    : %d\n",vol_base);
	    report("C4 replay rate : %d\n",c4_rate);
	    report("Channel mixing : %d%%\n",opt.mix*opt.reverse);
	}
	time(&t0);
	if(!play_module(d)) optind++;
	time(&t1);
	t=difftime(t1,t0);
	if(opt.verbose)
	    report("Elapsed time: %d:%02d\n",t/60,t%60);

	for(i=0;i<xxh->trk;i++) free(xxt[i]);
	for(i=0;i<xxh->pat;i++) free(xxp[i]);
	for(i=0;i<xxh->ins;i++) {
	    free(xxfe[i]);
	    free(xxpe[i]);
	    free(xxae[i]);
	    free(xxi[i]);
	}
	free(xxt);
	free(xxp);
	free(xxi);
	free(xxs);
	free(xxim);
	free(xxih);
	free(xxfe);
	free(xxpe);
	free(xxae);
    }
    return 0;
}

